### ISSUE TITLE
add Batch Normalization immediately before non-linearity or after in Keras?

### ISSUE THREAD

**Issue by daoliker** [fchollet/keras#5465]
def conv2d_bn(x, nb_filter, nb_row, nb_col,
                  border_mode='same', subsample=(1, 1),
                  name=None):
        '''Utility function to apply conv + BN.
        '''

        x = Convolution2D(nb_filter, nb_row, nb_col,
                          subsample=subsample,
                          activation='relu',
                          border_mode=border_mode,
                          name=conv_name)(x)
        x = BatchNormalization(axis=bn_axis, name=bn_name)(x)
        return x

When I use official [inception_v3][1] model in keras, I find that they use BatchNormalization after 'relu' nonlinearity as above code script.

But in the Batch Normalization paper, the authors said 

> we add the BN transform immediately before the nonlinearity, by
> normalizing x=Wu+b.

Then I view the implementation of inception in tensorflow which add BN immediately before the nonlinearity as they said. For more details in [inception ops.py](https://github.com/tensorflow/models/blob/master/inception/inception/slim/ops.py)

I'm confused. Why do people use above style in Keras other than the following? 

    def conv2d_bn(x, nb_filter, nb_row, nb_col,
                  border_mode='same', subsample=(1, 1),
                  name=None):
        '''Utility function to apply conv + BN.
        '''

        x = Convolution2D(nb_filter, nb_row, nb_col,
                          subsample=subsample,
                          border_mode=border_mode,
                          name=conv_name)(x)
        x = BatchNormalization(axis=bn_axis, name=bn_name)(x)
        x = Activation('relu')(x)
        return x

In the Dense case:

    x = Dense(1024, name='fc')(x)
    x = BatchNormalization(axis=bn_axis, name=bn_name)(x)
    x = Activation('relu')(x)

  [1]: https://github.com/fchollet/deep-learning-models/blob/master/inception_v3.py

**Comment by achalshah20** [comment]
Ideally, BN should be performed before non linearity. But I have seen some of the papers, where BN performed better if we apply after 'Relu'. 

Check this: https://github.com/ducha-aiki/caffenet-benchmark/blob/master/batchnorm.md

**Comment by Mickky666** [comment]
I'm reading the code of DCGAN, it seems that they used BN between the convolutional layer and lrelu. However, according to Jeremy Howard http://forums.fast.ai/t/questions-about-batch-normalization/230/2. He claims that we should use after non linearity.
